File: leetcode/1188-design-bounded-blocking-queue/solution.cpp
#include <condition_variable>
#include <deque>
#include <mutex>
using namespace std;

class BoundedBlockingQueue {
    int capacity;
    deque<int> queue;
    mutex mtx;
    condition_variable notFull;
    condition_variable notEmpty;

public:
    BoundedBlockingQueue(int capacity) : capacity(capacity) {}

    void enqueue(int element) {
        unique_lock<mutex> lock(mtx);
        notFull.wait(lock, [this] { return (int)queue.size() < capacity; });
        queue.push_back(element);
        notEmpty.notify_one();
    }

    int dequeue() {
        unique_lock<mutex> lock(mtx);
        notEmpty.wait(lock, [this] { return !queue.empty(); });
        int val = queue.front();
        queue.pop_front();
        notFull.notify_one();
        return val;
    }

    int size() {
        unique_lock<mutex> lock(mtx);
        return (int)queue.size();
    }
};

File: leetcode/1188-design-bounded-blocking-queue/Solution.java
import java.util.ArrayDeque;
import java.util.Deque;

class BoundedBlockingQueue {
    private final int capacity;
    private final Deque<Integer> queue = new ArrayDeque<>();
    private final Object lock = new Object();

    public BoundedBlockingQueue(int capacity) {
        this.capacity = capacity;
    }

    public void enqueue(int element) throws InterruptedException {
        synchronized (lock) {
            while (queue.size() == capacity) {
                lock.wait();
            }
            queue.addLast(element);
            lock.notifyAll();
        }
    }

    public int dequeue() throws InterruptedException {
        synchronized (lock) {
            while (queue.isEmpty()) {
                lock.wait();
            }
            int val = queue.removeFirst();
            lock.notifyAll();
            return val;
        }
    }

    public int size() {
        synchronized (lock) {
            return queue.size();
        }
    }
}
